baremetal/metal3data_manager.go

```diff
@@ -464,78 +490,87 @@ func (m *DataManager) m3IPClaimObjectMeta(name, poolRefName string, preallocatio
 	}
 }
 
-// getAddressFromM3Pool allocates an address from a metal3 IPAM pool using a IPClaim.
-// If the claim exists already, it will return ip address once the claim was fulfilled
-// with an IPAddress resource.
-func (m *DataManager) getAddressFromM3Pool(ctx context.Context, poolRef corev1.TypedLocalObjectReference) (addressFromPool, bool, error) {
+// ensureM3IPClaim ensures that a claim for a referenced pool exists.
+// It returns the claim and whether to fetch the claim again when fetching IP addresses.
+func (m *DataManager) ensureM3IPClaim(ctx context.Context, poolRef corev1.TypedLocalObjectReference) (reconciledClaim, error) {
 	ipClaim, err := fetchM3IPClaim(ctx, m.client, m.Log, m.Data.Name+"-"+poolRef.Name, m.Data.Namespace)
+	if err == nil {
+		return reconciledClaim{m3Claim: ipClaim}, nil
+	}
+
+	if ok := errors.As(err, &hasRequeueAfterError); !ok {
+		return reconciledClaim{m3Claim: ipClaim}, err
+	}
+
+	m3dt, err := fetchM3DataTemplate(ctx, &m.Data.Spec.Template, m.client,
+		m.Log, m.Data.Labels[clusterv1.ClusterLabelName],
+	)
 	if err != nil {
-		if ok := errors.As(err, &hasRequeueAfterError); !ok {
```

Comment:
the idea here was: we go fetch m3dt => m3m => bmh only, if IPClaim with normal "Data,Name+pool.Name" is not found, otherwise there is no need to have those objects to be fetched on each call. But now, we have to fetch them every time

Reply:
In line 497 I check whether err is nil, and return the claim if it is found. So the behaviour is still the same, just with less nesting.

Reply:
My bad, misread that